copy/copy.go

```diff
@@ -257,6 +297,10 @@ func (c *copier) copyOneImage(ctx context.Context, policyContext *signature.Poli
 		return nil, errors.Wrapf(err, "Error initializing image from source %s", transports.ImageName(c.rawSource.Reference()))
 	}
 
+	if options.EncryptLayers != nil && !src.SupportsEncryption(ctx) {
```

Comment:
Well… what _actually_ matters is whether encryption is supported by the manifest type that ends up being written to the destination (accounting for `otherManifestMIMETypeCandidates`).

This might end up being good enough — or it might be a suggestion that the whole capability/edit process needs a deeper look.

(There are strong parallels to the compression decisions, which have not dealt with this correctly I think.)

Reply:
I agree with this! I think the challenge that we ran into here was the way [`(ic *imageCopier) copyUpdatedConfigAndManifest`](https://github.com/containers/image/blob/master/copy/copy.go#L558) worked. I'm guessing we need to rewrite that function to copy to dest then update instead of the other way around.

Reply:
Well… the layers must be uploaded before it’s valid to upload a manifest, and it’s only when uploading the manifest that we can tell whether the registry supports that manifest type or not.

So, it’s not something the code can always predict correctly, without possibly encrypting all that data and then failing; therefore, it _must_ be fine for the encryption decision to happen before that `otherManifestMIMETypeCandidates` loop, and for that encryption decision to restrict which of the other candidates have a chance of succeeding.

It’s more a matter of correctly dealing with `copy.Options.ForceManifestMIMEType` right now.

Reply:
(To an extent, it’s quite fine to leave some of the complex cases unimplemented, if they clearly fail with an error. I’m more worried about embedding the incorrect assumptions into the stable API.)

Reply:
I've taken a stab at what could be done to avoid this - minus the need to change the way the actual manifest updates are done still on the source than the destination. 

What do you think of something like this?

https://github.com/containers/image/pull/673/commits/78c7bece4c6f6e977aec63b832b0d0adf2e6e9b4

Reply:
If this is something that works, I can look into how we can move the manifest update to be based on destination manifest instead so we can remove that source mimetype dependency.

Reply:
@mtrmac Any thoughts on this idea?

Reply:
(Looking at the linked commit https://github.com/containers/image/commit/78c7bece4c6f6e977aec63b832b0d0adf2e6e9b4 ):

- There is `determineManifestConversion`, `supportsEncryption` should not duplicate (insufficiently) its logic
- `manifestSupportsEncryption` looks like something that should be provided by `c/image/manifest` (directly by the manifest type, or in some other abstraction?), not hard-coded in `c/image/copy`
  - (The same, to an extent, applies to `isOCIEncrypted`)

Reply:
- Is the “TODO: Remove src.SupportsEncryption call…” item supposed to stay there, or should that already have been removed?

Reply:
> It can't be removed right now because of the way manifests are updated, which use the source manifest instead of destination manifest.

Ah, right. If that isn’t fixed before this PR, the comment should at least clearly explain that this is the reason the check needs to exist.

Reply:
> > There is determineManifestConversion, supportsEncryption should not duplicate (insufficiently) its logic
> 
> This would mean that we would have to pass in `src` image to `determineManifestConversion`.

(Discussed separately — that’s not what the concern was, but moving all of the logic to `determineManifestConversion` might cause `supportsEncryption` to be unnecessary anyway.)

> I initial thought that we don't want to create public interfaces. Does this change the initial concern? [#673 (comment)](https://github.com/containers/image/pull/673#discussion_r332082882)

Public APIs are a long-term commitment (or, so far, a reason to very frequently bump major versions, which is something I’d very much prefer to stop, and why I’m so worried about them). At the time of that comment it seemed that the function could be private, so making it private would have easily made the concern moot and would have given us complete flexibility to redesign that later as necessary.

If we do need a public function, or if avoiding a public function would lead to worse design (in this case tightly coupling c/image/copy and c/image/manifest) and breaking abstractions), the need to commit to it is not a _show-stopper_, but it’s a good reason to try hard to make the design general and long-lasting (which is why https://github.com/containers/image/pull/673#discussion_r332082882 tentatively suggests making it a per-format method of `c/image/manifest.Manifest` instead of a global function that understands all formats).

The “manifestSupportsEncryption” question is a bad match for `c/image/manifest.Manifest`, so maybe we need a `c/image/manifest.ManifestFormat`? No idea. Or maybe having a `MIMETypeSupportsEncryption` function very similar to the `manifestSupportsEncryption` one would work well enough (there’s at least one precedent for that already, and all of this somewhat hard-codes the _OCI_ encryption format anyway).